### PR TITLE
[pt1][quant] Add the type matching rule for qconfig_dict

### PR DESCRIPTION
**Pull request by jianyuh**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23139 [pt1][quant] Unit Test for the Dynamic Quantized Linear operator
* #24231 [pt1][quant] Add the default_weight_observer for the dynamic quantization path
* #24299 [pt1][quant] Remove the activation observer for default_qconfig
* #24232 [pt1][quant] Make the default qconfig_dict
* #23330 [pt1][quant] Fix the dimension mismatch issues when running the BERT model
* **#23212 [pt1][quant] Add the type matching rule for qconfig_dict**

We want to use the Module type as the key for the qconfig_dict for the module replacement during the quantization.

Before this Diff, to dynamic quantize the BERT model, we have to specify each layer:
```
qconfig_dict = {
    'encoder.layer.0.attention.self.query': default_qconfig,
    'encoder.layer.0.attention.self.key': default_qconfig,
    'encoder.layer.0.attention.self.value': default_qconfig,
    'encoder.layer.0.attention.output.dense': default_qconfig,
    'encoder.layer.0.intermediate.dense': default_qconfig,
    'encoder.layer.0.output.dense': default_qconfig,
    'encoder.layer.1.attention.self.query': default_qconfig,
    'encoder.layer.1.attention.self.key': default_qconfig,
    'encoder.layer.1.attention.self.value': default_qconfig,
    'encoder.layer.1.attention.output.dense': default_qconfig,
    'encoder.layer.1.intermediate.dense': default_qconfig,
    'encoder.layer.1.output.dense': default_qconfig,
   ...
}
```
After this Diff, we only need the following
```
qconfig_dict = {
     torch.nn.Linear : default_qconfig
}
```

Differential Revision: [D16436542](https://our.internmc.facebook.com/intern/diff/D16436542/)

### CODE REVIEW

File: torch/quantization/quantize.py [hunk 1]
```diff
@@ -23,8 +23,13 @@ def propagate_qconfig_helper(module, qconfig_dict, qconfig_parent=None, prefix='
...
     if not hasattr(module, 'qconfig'):
         module.qconfig = None
-        if qconfig_dict and prefix in qconfig_dict:
-            module.qconfig = qconfig_dict[prefix]
+        if qconfig_dict:
+            if prefix in qconfig_dict:
+                module.qconfig = qconfig_dict[prefix]
+            elif type(module) in qconfig_dict:
+                module.qconfig = qconfig_dict[type(module)]
+            else:
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: I think we don't need this, we can have `module.qconfig=qconfig_parent` in line 25 and remove all the else branches.

Comment (jianyuh) [reply torch/quantization/quantize.py]: Updated! Thanks @jerryzh168 !

File: test/test_quantization.py [hunk 2]
```diff
@@ -420,6 +420,39 @@ def checkQuantized(model):
         model = quantize_dynamic(NestedModel().eval(), test_only_eval_fn, self.calib_data, qconfig_dict)
         checkQuantized(model)
 
+    def test_type_match_rule(self):
+        r"""Test quantization for nested model, top level 'fc3' and
+        'fc1' of submodule 'sub2', All 'torch.nn.Linear' modules are quantized
+        """
+        model = NestedModel().eval()
+        qconfig_dict = {
+            'fc3': default_qconfig,
```
Comment (raghuramank100) [inline test/test_quantization.py]: Is it possible to have an example where:
qconfig_dict = {
            'fc3': None, 
            torch.nn.Linear:default_qconfig
}
and show that all layers except fc3 are dynamically quantized?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]

File: test/test_quantization.py (hunk 2 above)
Comment (jianyuh) [reply test/test_quantization.py]: Updated. In the new version, we choose not to quantize `fc3` and `sub2.fc1`. We show `sub1.fc` and `sub2.fc2` are dynamic quantized, while `fc3` and `sub2.fc1` remain the same.

**Review by z-a-f — APPROVED** [review]
LGTM

**Review by raghuramank100 — APPROVED** [review]

**Pull request closed without merging**